### PR TITLE
tools/renode: refactor integration + add debugger support

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR is a complete refactoring of the renode emulator integration. It's based on #15486 and provides several enhancements:
- the basic renode startup code is moved to Makefile only code by using a series of `-e` command instead of a temporary config file generated on the fly. As a result the run-renode.sh script is removed.
- integration of debug and debug-server targets is added: it's now possible to call `emulate debug` targets together to start renode with a gdb server and to automatically attached to it using gdb.
When using the `debug%` target with `emulate` silents completely `emulate`: in the proposed approach, we can't have the renode GDB server running at the same time as the basic emulator. This also allows to reuse the existing debug% targets as-is and doesn't conflicts with existing debuggers (without emulate, the default on-chip debugger is used).
- renode `start` command is removed from the renode board configuration of the cc2538dk board (only board supported so far): this is because this is not working when using the debugger (as this ones needs a different start sequence). The renode `start` command is just called using another interactive command defined in `renode.inc.mk` and called at startup with the `emulate` target alone
- the `emulate` target now depends on `EMULATORDEPS` and all: this is much more flexible than just depend on `all` since it allows us to add other dependencies from the board configuration or the different emulators configuration. For example, `EMULATORDEPS` is used by renode to add the board configuration (.resc) as a dependency. So trying to emulate a board without a renode config will fail early because of missing dependency
- an `emulate-only` target is also added and doesn't depend on `all`. This is especially useful when using `debug`.

### Testing procedure

- Check that `emulate` still works and triggers a rebuild: `make BOARD=cc2538dk -C examples/hello-world emulate`

<details><summary>emulate alone</summary>

```
$ BUILD_IN_DOCKER=1 make BOARD=cc2538dk -C examples/hello-world emulate --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=cc2538dk'  -w '/data/riotbuild/riotbase/examples/hello-world/' 'riot/riotbuild:latest' make 'BOARD=cc2538dk'    
Building application "hello-world" for "cc2538dk" with MCU "cc2538".

"make" -C /data/riotbuild/riotbase/boards/cc2538dk
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/cc2538
"make" -C /data/riotbuild/riotbase/cpu/cc2538/periph
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/pm_layered
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
   text	   data	    bss	    dec	    hex	filename
   8856	    112	   2296	  11264	   2c00	/data/riotbuild/riotbase/examples/hello-world/bin/cc2538dk/hello-world.elf
renode -e "set image_file '/work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf'" -e "include @/work/riot/RIOT/boards/cc2538dk/dist/board.resc" -e start
Gtk-Message: 19:07:48.849: Failed to load module "canberra-gtk-module"
19:07:50.1634 [INFO] Loaded monitor commands from: /opt/renode/./scripts/monitor.py
19:07:50.5459 [INFO] Including script: /work/riot/RIOT/boards/cc2538dk/dist/board.resc
19:07:50.5916 [INFO] System bus created.
19:07:51.8833 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00020000, 0x00031FFF>.
19:07:51.8834 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00032000, 0x0004AFFF>.
19:07:51.8836 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x0004B000, 0x00063FFF>.
19:07:51.8837 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00064000, 0x0007CFFF>.
19:07:51.8838 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x0007D000, 0x0007FFFF>.
19:07:51.8934 [INFO] sysbus: Loading segment of 8812 bytes length at 0x200000.
19:07:51.8938 [INFO] sysbus: Loading segment of 1896 bytes length at 0x20226C.
19:07:51.8938 [INFO] sysbus: Loading segment of 44 bytes length at 0x27FFD4.
19:07:51.8939 [INFO] sysbus: Loading segment of 512 bytes length at 0x20000000.
19:07:52.1214 [INFO] cpu: Setting initial values: PC = 0x200831, SP = 0x20000200.
19:07:52.1235 [INFO] machine-0: Machine started.
```

<p align=center>

</p>

</details>

- Check that `emulate debug` start the GDB server in renode and the gdb client automatically attach to it: `make BOARD=cc2538dk -C examples/hello-world emulate debug`

<details><summary>emulate debug</summary>

```
$ make BOARD=cc2538dk -C examples/hello-world emulate debug --no-print-directory 
Building application "hello-world" for "cc2538dk" with MCU "cc2538".

"make" -C /work/riot/RIOT/boards/cc2538dk
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/cc2538
"make" -C /work/riot/RIOT/cpu/cc2538/periph
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
   text	   data	    bss	    dec	    hex	filename
   8852	    112	   2296	  11260	   2bfc	/work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf
/work/riot/RIOT/dist/tools/renode/renode-debug.sh cc2538dk /work/riot/RIOT/examples/hello-world /work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf 3333
Using Renode pid file /tmp/renode_debug_server_pid.RgKhH8HY5I
Reading symbols from /work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf...
renode -e "set image_file '/work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf'" -e "include @/work/riot/RIOT/boards/cc2538dk/dist/board.resc" -e "machine StartGdbServer 3333 true" --pid-file /tmp/renode_debug_server_pid.RgKhH8HY5I
Gtk-Message: 19:19:58.514: Failed to load module "canberra-gtk-module"
19:19:59.7794 [INFO] Loaded monitor commands from: /opt/renode/./scripts/monitor.py
19:20:00.2819 [INFO] Including script: /work/riot/RIOT/boards/cc2538dk/dist/board.resc
19:20:00.3209 [INFO] System bus created.
19:20:01.5678 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00020000, 0x00031FFF>.
19:20:01.5679 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00032000, 0x0004AFFF>.
19:20:01.5680 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x0004B000, 0x00063FFF>.
19:20:01.5681 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00064000, 0x0007CFFF>.
19:20:01.5682 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x0007D000, 0x0007FFFF>.
19:20:01.5773 [INFO] sysbus: Loading segment of 8808 bytes length at 0x200000.
19:20:01.5776 [INFO] sysbus: Loading segment of 1896 bytes length at 0x202268.
19:20:01.5777 [INFO] sysbus: Loading segment of 44 bytes length at 0x27FFD4.
19:20:01.5778 [INFO] sysbus: Loading segment of 512 bytes length at 0x20000000.
19:20:01.8257 [INFO] machine-0: GDB server with all CPUs started on port :3333
Remote debugging using :3333
19:20:04.6867 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFF>.
19:20:04.7300 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFF>.
19:20:04.7318 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFE, 0xFFFFFFFF>.
19:20:04.7329 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFD>.
19:20:04.7346 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFE, 0xFFFFFFFF>.
19:20:04.7356 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFD>.
19:20:04.7373 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFF>.
19:20:04.7826 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFF>.
0x00000000 in ?? ()
19:20:04.8596 [INFO] cpu: Setting initial values: PC = 0x200831, SP = 0x20000200.
19:20:04.8706 [INFO] machine-0: Machine started.
Starting emulation...
(gdb) load
Loading section .text, size 0x2268 lma 0x200000
Loading section .relocate, size 0x70 lma 0x202268
Loading section .flashcca, size 0x2c lma 0x27ffd4
Start address 0x00200830, load size 8964
Transfer rate: 71 KB/sec, 1792 bytes/write.
(gdb) b main
Breakpoint 1 at 0x20028c: file /work/riot/RIOT/examples/hello-world/main.c, line 26.
(gdb) c
Continuing.
19:20:16.7277 [WARNING] nvic: Unhandled read from offset 0xD94.
19:20:16.7294 [WARNING] nvic: Unhandled write to offset 0xD94, value 0x5.
19:20:16.7386 [WARNING] nvic: Unhandled value written to System Control Register: 0x10.
19:20:16.7387 [WARNING] nvic: Unhandled read from offset 0xD14.
19:20:16.7387 [WARNING] nvic: Unhandled write to offset 0xD14, value 0x200.
19:20:16.7422 [WARNING] sysbus: [cpu: 0x200A7A] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2098, value 0x1.
19:20:16.7497 [WARNING] sysbus: [cpu: 0x200A7A] WriteDoubleWord to non existing peripheral at 0x400D40F8, value 0x1.
19:20:16.7499 [WARNING] sysbus: [cpu: 0x200A7A] WriteDoubleWord to non existing peripheral at 0x400D40FC, value 0x1.
19:20:16.7503 [WARNING] sysbus: [cpu: 0x200A7A] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.7503 [WARNING] sysbus: [cpu: 0x200A7A] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x10001.
19:20:16.8401 [WARNING] sysbus: [cpu: 0x200AD0] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8402 [WARNING] sysbus: [cpu: 0x200AD0] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x1.
19:20:16.8507 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8508 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x1.
19:20:16.8508 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8508 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x20001.
19:20:16.8509 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8509 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x0.
19:20:16.8509 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8510 [WARNING] sysbus: [cpu: 0x200AE4] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x101.
19:20:16.8784 [WARNING] sysbus: [cpu: 0x200C40] WriteDoubleWord to non existing peripheral at 0x400D4084, value 0x8.
19:20:16.8785 [WARNING] sysbus: [cpu: 0x200C40] WriteDoubleWord to non existing peripheral at 0x400D4004, value 0x0.
19:20:16.8897 [WARNING] gpioA: Unhandled write to offset 0x420. Unhandled bits: [1] when writing value 0x2. Tags: AFSEL (0x2).
19:20:16.8898 [WARNING] sysbus: [cpu: 0x200D50] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2028, returning 0x1.
19:20:16.8900 [WARNING] sysbus: [cpu: 0x200D50] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2028, value 0x1.
19:20:16.8900 [WARNING] sysbus: [cpu: 0x200D50] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D202C, returning 0x1.
19:20:16.8901 [WARNING] sysbus: [cpu: 0x200D50] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D202C, value 0x1.
19:20:16.8901 [WARNING] sysbus: [cpu: 0x200D50] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2030, returning 0x1.
19:20:16.8902 [WARNING] sysbus: [cpu: 0x200D50] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2030, value 0x1.
19:20:16.8918 [WARNING] uart0: Unhandled write to offset 0xFC8, value 0x0.
19:20:16.8930 [WARNING] sysbus: [cpu: 0x200D8E] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8931 [WARNING] sysbus: [cpu: 0x200D8E] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
19:20:16.8954 [WARNING] sysbus: [cpu: 0x200C00] WriteDoubleWord to non existing peripheral at 0x400D40C0, value 0x8.
19:20:16.8969 [WARNING] sysbus: [cpu: 0x200C00] WriteDoubleWord to non existing peripheral at 0x400D40C4, value 0x8.
19:20:16.8973 [WARNING] sysbus: [cpu: 0x200C00] WriteDoubleWord to non existing peripheral at 0x400D40C8, value 0x8.
19:20:16.8978 [WARNING] sysbus: [cpu: 0x200C00] WriteDoubleWord to non existing peripheral at 0x400D40CC, value 0x8.
19:20:16.9077 [WARNING] nvic: Unhandled write to offset 0xD98, value 0x2.
19:20:16.9078 [WARNING] nvic: Unhandled write to offset 0xD9C, value 0x20000280.
19:20:16.9079 [WARNING] nvic: Unhandled write to offset 0xDA0, value 0x16060009.

Breakpoint 1, main () at /work/riot/RIOT/examples/hello-world/main.c:26
26	    puts("Hello World!");
(gdb) l
21	
22	#include <stdio.h>
23	
24	int main(void)
25	{
26	    puts("Hello World!");
27	
28	    printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
29	    printf("This board features a(n) %s MCU.\n", RIOT_MCU);
30	
(gdb) quit
A debugging session is active.

	Inferior 1 [Remote target] will be killed.

Quit anyway? (y or n) EOF [assumed Y]
Terminated
```

</details>

- Check that `emulate debug-server` start the GDB server in renodet: `make BOARD=cc2538dk -C examples/hello-world emulate debug-server`. It's then possible to attach to the GDB server using `gdb <elffile> -q -ex "target remote :3333" -ex "monitor start"`

<details><summary>emulate debug-server</summary>

```
$ make BOARD=cc2538dk -C examples/hello-world emulate debug-server --no-print-directory 
Building application "hello-world" for "cc2538dk" with MCU "cc2538".

"make" -C /work/riot/RIOT/boards/cc2538dk
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/cc2538
"make" -C /work/riot/RIOT/cpu/cc2538/periph
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
   text	   data	    bss	    dec	    hex	filename
   8852	    112	   2296	  11260	   2bfc	/work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf
renode -e "set image_file '/work/riot/RIOT/examples/hello-world/bin/cc2538dk/hello-world.elf'" -e "include @/work/riot/RIOT/boards/cc2538dk/dist/board.resc" -e "machine StartGdbServer 3333 true"
Gtk-Message: 19:25:45.352: Failed to load module "canberra-gtk-module"
19:25:48.4800 [INFO] Loaded monitor commands from: /opt/renode/./scripts/monitor.py
19:25:49.1069 [INFO] Including script: /work/riot/RIOT/boards/cc2538dk/dist/board.resc
19:25:49.1543 [INFO] System bus created.
19:25:51.1251 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00020000, 0x00031FFF>.
19:25:51.1253 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00032000, 0x0004AFFF>.
19:25:51.1255 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x0004B000, 0x00063FFF>.
19:25:51.1256 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x00064000, 0x0007CFFF>.
19:25:51.1259 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0x0007D000, 0x0007FFFF>.
19:25:51.1463 [INFO] sysbus: Loading segment of 8808 bytes length at 0x200000.
19:25:51.1471 [INFO] sysbus: Loading segment of 1896 bytes length at 0x202268.
19:25:51.1472 [INFO] sysbus: Loading segment of 44 bytes length at 0x27FFD4.
19:25:51.1473 [INFO] sysbus: Loading segment of 512 bytes length at 0x20000000.
19:25:51.5240 [INFO] machine-0: GDB server with all CPUs started on port :3333
```

in another terminal:
```
$ gdb-multiarch /work/riot/RIOT/examples/default/bin/cc2538dk/default.elf -ex "target remote :3333" -ex "monitor start"
GNU gdb (Ubuntu 9.2-0ubuntu2) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /work/riot/RIOT/examples/default/bin/cc2538dk/default.elf...
Remote debugging using :3333
0x00000000 in ?? ()
Starting emulation...
(gdb) load
Loading section .text, size 0x9040 lma 0x200000
Loading section .relocate, size 0x84 lma 0x209040
Loading section .flashcca, size 0x2c lma 0x27ffd4
Start address 0x002017d0, load size 37104
Transfer rate: 180 KB/sec, 3092 bytes/write.
(gdb) b main
Breakpoint 1 at 0x20028c: file /work/riot/RIOT/examples/default/main.c, line 40.
(gdb) l
26	#include <string.h>
27	
28	#include "thread.h"
29	#include "shell.h"
30	#include "shell_commands.h"
31	
32	#ifdef MODULE_NETIF
33	#include "net/gnrc/pktdump.h"
34	#include "net/gnrc.h"
35	#endif
(gdb) c
Continuing.

Breakpoint 1, main () at /work/riot/RIOT/examples/default/main.c:40
40	    gnrc_netreg_entry_t dump = GNRC_NETREG_ENTRY_INIT_PID(GNRC_NETREG_DEMUX_CTX_ALL,
(gdb) l
35	#endif
36	
37	int main(void)
38	{
39	#ifdef MODULE_NETIF
40	    gnrc_netreg_entry_t dump = GNRC_NETREG_ENTRY_INIT_PID(GNRC_NETREG_DEMUX_CTX_ALL,
41	                                                          gnrc_pktdump_pid);
42	    gnrc_netreg_register(GNRC_NETTYPE_UNDEF, &dump);
43	#endif
44	
(gdb) quit
```

</details>

### Issues/PRs references

closes #15483 (since exports are also removed from this PR)
based on #15486 and #15475

### CODE REVIEW

**Comment by aabadie** [comment]
I found some interaction problems between this PR and existing programmers so I had to rebase on top of #15475 and provide some adaptation: in master the programmer management code is included unconditionally and programmers also define the `DEBUGGER*`/`DEBUGSERVER*` that are used by the renode emulator. And since programmer makefile is included before emulator makefile code, the variable are already defined.
With cc2538dk this is not really a problem because the cc2538-bsl doesn't provide these variables but for other boards (stm32f4discovery for example) `emulate debug` won't work.
This PR also now contains 4ed2215 because emulator makefile is included before programmer makefiles and the `DEBUGGER*`/`DEBUGSERVER*` shouldn't be redefined by programmers if already defined by the emulator.

In the end, there are lots of changes required in the build system to be able to have a nice integration of `emulate debug`. But I think they are useful and are enforcing cleaner things.

**Comment by aabadie** [comment]
FYI, I have more to come: integration of `term` with renode. This would allow to use the `test` target with renode as well.

**Comment by aabadie** [comment]
rebased and no longer waiting for other PRs!

File: boards/cc2538dk/dist/board.resc [hunk 1]
```diff
@@ -27,4 +27,3 @@ macro reset
 """
 
 runMacro $reset
-start
```
Comment (fjmolinas) [inline boards/cc2538dk/dist/board.resc:30]: Why this change?

**Review by fjmolinas — COMMENTED** [review]
Looks good in general, will need to properly test this. I would like to brainstorm about how to handle the `emulate` commands.  Options I can think of are

- have a environment variable like for `BUILD_IN_DOCKER` and have all targets use the emulator (all possible targets). e.g.: `EMULATE=1 make term`
- use something similar as #12419 and have `emualate/%` targets, so `emulate/term` `emulate/flash`.

Only having an `emulate` target would be weirder to handler.

File: boards/cc2538dk/dist/board.resc (hunk 1 above)
Comment (aabadie) [reply boards/cc2538dk/dist/board.resc:30]: because that doesn't work with the debugger otherwise. The startup sequence in the monitor is different. When using `emulate` alone, the start command is called using the `-e star` option of renode directly. When using `emulate debug*` targets, other commands are used to start the debug server and automatically start the emulated machine. If start is called before (from the .resc file), that doesn't work, the debug server is stuck and the gdb client cannot attach to it.

Also note that since #15501 and #15489 are merged, I have to adapt their configuration files as well.

Comment (aabadie) [reply boards/cc2538dk/dist/board.resc:30]: > Also note that since #15501 and #15489 are merged, I have to adapt their configuration files as well.

PR adapted

**Comment by aabadie** [comment]
> have a environment variable like for BUILD_IN_DOCKER and have all targets use the emulator (all possible targets). e.g.: EMULATE=1 make term

This is done by #15512 

> use something similar as #12419 and have emualate/% targets, so emulate/term emulate/flash

The drawback of this solution is that it prevent a transparent usage of the `test` target that is using `make term` internally. With the first solution you get it for free.

**Comment by fjmolinas** [comment]
> > have a environment variable like for BUILD_IN_DOCKER and have all targets use the emulator (all possible targets). e.g.: EMULATE=1 make term
> 
> This is done by #15512
> 
> > use something similar as #12419 and have emualate/% targets, so emulate/term emulate/flash
> 
> The drawback of this solution is that it prevent a transparent usage of the `test` target that is using `make term` internally. With the first solution you get it for free.

Ok lets discuss in that PR then.

**Comment by fjmolinas** [comment]
When calling `make emulate debug`: the renode logs gets mixed up with the gdb console :/

```
17:22:26.6332 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFF>.
17:22:26.6772 [WARNING] sysbus: Tried to access bytes at non-existing peripheral in range <0xFFFFFFFC, 0xFFFFFFFF>.
0x00000000 in ?? ()
17:22:26.6974 [INFO] cpu: Setting initial values: PC = 0x200839, SP = 0x20000200.
17:22:26.7007 [INFO] machine-0: Machine started.
Starting emulation...
(gdb) b main
Breakpoint 1 at 0x20028c: file /home/francisco/workspace/RIOT2/examples/hello-world/main.c, line 26.
(gdb) r
The "remote" target does not support "run".  Try "help target" or "continue".
(gdb) continue
Continuing.
17:22:57.5045 [WARNING] nvic: Unhandled read from offset 0xD94.
17:22:57.5063 [WARNING] nvic: Unhandled write to offset 0xD94, value 0x5.
17:22:57.5147 [WARNING] nvic: Unhandled value written to System Control Register: 0x10.
17:22:57.5147 [WARNING] nvic: Unhandled read from offset 0xD14.
17:22:57.5147 [WARNING] nvic: Unhandled write to offset 0xD14, value 0x200.
17:22:57.5192 [WARNING] sysbus: [cpu: 0x200A82] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2098, value 0x1.
17:22:57.5305 [WARNING] sysbus: [cpu: 0x200A82] WriteDoubleWord to non existing peripheral at 0x400D40F8, value 0x1.
17:22:57.5306 [WARNING] sysbus: [cpu: 0x200A82] WriteDoubleWord to non existing peripheral at 0x400D40FC, value 0x1.
17:22:57.5310 [WARNING] sysbus: [cpu: 0x200A82] (tag: 'SYS_CTRL_CLOCK_CTRL') ReadDoubleWord from non existing peripheral at 0x400D2000, returning 0x1.
17:22:57.5311 [WARNING] sysbus: [cpu: 0x200A82] (tag: 'SYS_CTRL_CLOCK_CTRL') WriteDoubleWord to non existing peripheral at 0x400D2000, value 0x10001.

```

**Comment by aabadie** [comment]
> When calling make emulate debug: the renode logs gets mixed up with the gdb console 

Does the renode window appear as well ? (I don't remember ;) )

**Comment by fjmolinas** [comment]
> Does the renode window appear as well ? (I don't remember ;) )

Yes, I get three windows.

**Comment by aabadie** [comment]
> I get three windows

3 ? Normally, there's just the renode monitor + serial output. What is the third one ?

**Comment by fjmolinas** [comment]
> > I get three windows
> 
> 3 ? Normally, there's just the renode monitor + serial output. What is the third one ?

The third is the terminal :)

**Comment by aabadie** [comment]
Renode options are quite limited. We could completely silent (or at least make it configurable) the log output. This is done in #15512

**Comment by fjmolinas** [comment]
> Renode options are quite limited. We could completely silent (or at least make it configurable) the log output. This is done in #15512

Ok, then addressed in a follow up. `debug-server` works as expected.

**Review by fjmolinas — APPROVED** [review]
ACK, not happy with every detail in the usability, but I takes this as a first step. And #15512 seems to address most of the expressed concerns
